### PR TITLE
partition: fix partitioning pruning includes an impossible partition

### LINKED ISSUE #42273: RANGE COLUMNS partitioning pruning includes an impossible partition
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE t
(a tinyint unsigned,
 b tinyint unsigned)
PARTITION BY RANGE COLUMNS (a,b)
(PARTITION p0 VALUES LESS THAN (10,255),
 PARTITION p1 VALUES LESS THAN (11,MAXVALUE),
 PARTITION p2 VALUES LESS THAN (MAXVALUE,1));
explain select * from t where a = 11;
analyze table t; -- Enables dynamic prune mode
explain select * from t where a = 11;
```

### 2. What did you expect to see? (Required)
Only partition p1 should be included, like MySQL:
```
mysql> CREATE TABLE t
    -> (a tinyint unsigned,
    ->  b tinyint unsigned)
    -> PARTITION BY RANGE COLUMNS (a,b)
    -> (PARTITION p0 VALUES LESS THAN (10,255),
    ->  PARTITION p1 VALUES LESS THAN (11,MAXVALUE),
    ->  PARTITION p2 VALUES LESS THAN (MAXVALUE,1));
Query OK, 0 rows affected (0,03 sec)

mysql> explain select * from t where a = 11;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | p1         | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0,00 sec)
```

### 3. What did you see instead (Required)
Partition p2 is included, when it cannot hold any values where a = 11.

```
tidb> CREATE TABLE t
    -> (a tinyint unsigned,
    ->  b tinyint unsigned)
    -> PARTITION BY RANGE COLUMNS (a,b)
    -> (PARTITION p0 VALUES LESS THAN (10,255),
    ->  PARTITION p1 VALUES LESS THAN (11,MAXVALUE),
    ->  PARTITION p2 VALUES LESS THAN (MAXVALUE,1));
Query OK, 0 rows affected (0.08 sec)

tidb> explain select * from t where a = 11;
+-----------------------------+----------+-----------+-----------------------+--------------------------------+
| id                          | estRows  | task      | access object         | operator info                  |
+-----------------------------+----------+-----------+-----------------------+--------------------------------+
| PartitionUnion_8            | 20.00    | root      |                       |                                |
| ├─TableReader_11        | 10.00    | root      |                       | data:Selection_10              |
| │ └─Selection_10      | 10.00    | cop[tikv] |                       | eq(test.t.a, 11)               |
| │   └─TableFullScan_9 | 10000.00 | cop[tikv] | table:t, partition:p1 | keep order:false, stats:pseudo |
| └─TableReader_14        | 10.00    | root      |                       | data:Selection_13              |
|   └─Selection_13        | 10.00    | cop[tikv] |                       | eq(test.t.a, 11)               |
|     └─TableFullScan_12  | 10000.00 | cop[tikv] | table:t, partition:p2 | keep order:false, stats:pseudo |
+-----------------------------+----------+-----------+-----------------------+--------------------------------+
7 rows in set, 1 warning (0.00 sec)

tidb> analyze table t; -- Enables dynamic prune mode
Query OK, 0 rows affected, 6 warnings (0.15 sec)

tidb> explain select * from t where a = 11;
+-------------------------+----------+-----------+-----------------+--------------------------------+
| id                      | estRows  | task      | access object   | operator info                  |
+-------------------------+----------+-----------+-----------------+--------------------------------+
| TableReader_7           | 10.00    | root      | partition:p1,p2 | data:Selection_6               |
| └─Selection_6       | 10.00    | cop[tikv] |                 | eq(test.t.a, 11)               |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t         | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+-----------------+--------------------------------+
3 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v6.6.0
Edition: Community
Git Commit Hash: f4ca0821fb96a2bdd37d2fb97eb26c07fc58d4e4
Git Branch: heads/refs/tags/v6.6.0
UTC Build Time: 2023-02-17 14:39:38
GoVersion: go1.19.5
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by jiyfhust**
### What problem does this PR solve?

Issue Number: close #42273

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -1145,6 +1145,13 @@ func maxCmp(ctx sessionctx.Context, hiVal []types.Datum, columnsPruner *rangeCol
 				return false
 			}
 		}
+		if len(hiVal) < len(columnsPruner.lessThan[i]) {
```
Comment (mjonss) [inline planner/core/rule_partition_processor.go:1148]:
```suggestion
		// All hiVal == columnsPruner.lessThan
		if len(hiVal) < len(columnsPruner.lessThan[i]) {
```

File: planner/core/partition_pruner_test.go [hunk 2]
```diff
@@ -453,3 +453,27 @@ func TestIssue33231(t *testing.T) {
...
 }
+
+func TestIssue42273(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("create database issue42273")
+	defer tk.MustExec("drop database issue42273")
+
+	tk.MustExec("use issue42273")
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'dynamic';")
```
Comment (mjonss) [inline planner/core/partition_pruner_test.go:464]:
```suggestion
```
This should already be the default

File: planner/core/partition_pruner_test.go [hunk 3]
```diff
@@ -453,3 +453,27 @@ func TestIssue33231(t *testing.T) {
...
+	tk.MustExec(`CREATE TABLE t(a tinyint unsigned,  b tinyint unsigned) PARTITION BY RANGE COLUMNS (a,b)(
+			PARTITION p0 VALUES LESS THAN (10,255),
+			PARTITION p1 VALUES LESS THAN (20,MAXVALUE),
+			PARTITION p2 VALUES LESS THAN (30,255),
+			PARTITION p3 VALUES LESS THAN (MAXVALUE, 0))`)
+	tk.MustQuery(`explain format='brief' select * from t where a = 20`).Check(testkit.Rows("TableReader 10.00 root  data:Selection",
```
Comment (mjonss) [inline planner/core/partition_pruner_test.go:470]:
```suggestion
	tk.MustExec("insert into t values(20, 30)")
	tk.MustExec(`analyze table t`) // Creates global stats for the table and enables the dynamic pruning
	tk.MustQuery(`explain format='brief' select * from t where a = 20`).Check(testkit.Rows("TableReader 10.00 root  data:Selection",
```

File: planner/core/partition_pruner_test.go [hunk 4]
```diff
@@ -453,3 +453,27 @@ func TestIssue33231(t *testing.T) {
...
+		"└─Selection 10.00 cop[tikv]  eq(issue42273.t.a, 20)",
+		"  └─TableFullScan 10000.00 cop[tikv] table:t, partition:p1 keep order:false, stats:pseudo"))
+	tk.MustQuery(`explain format='brief' select * from t where a > 10 and a <= 20`).Check(testkit.Rows("TableReader 250.00 root  data:Selection",
+		"└─Selection 250.00 cop[tikv]  gt(issue42273.t.a, 10), le(issue42273.t.a, 20)",
+		"  └─TableFullScan 10000.00 cop[tikv] table:t, partition:p1 keep order:false, stats:pseudo"))
+	tk.MustExec("insert into t values(20, 30)")
```
Comment (mjonss) [inline planner/core/partition_pruner_test.go:476]:
```suggestion
```

**Review by mjonss — APPROVED** [review]
Great catch! LGTM, just some suggestions for the test, currently it actually is using static pruning mode, due to an issue when there is no global statistics, it falls back to static prune mode.

**Comment by mjonss** [comment]
@jiyfhust please check and update the tests TestRangeColumnsExpr and TestRangeMultiColumnsPruning

**Comment by jiyfhust** [comment]
> @jiyfhust please check and update the tests TestRangeColumnsExpr and TestRangeMultiColumnsPruning

@mjonss done.

**Comment by jiyfhust** [comment]
```
--- PASS: TestLargeVarcharAutoConv (0.92s)

FAIL

coverage: 18.9% of statements

--

Coverage runner: Not collecting coverage for failed test.

The following commands failed with status 1

/home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/9338/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/core/core_test_/core_test.runfiles/__main__/planner/core/core_test_/core_test

================================================================================

ERROR: Test failed, aborting

INFO: Elapsed time: 484.343s, Critical Path: 326.50s

INFO: 10078 processes: 6345 remote cache hit, 1167 internal, 2566 linux-sandbox.

INFO: Build completed, 1 test FAILED, 10078 total actions

Test cases: finished with 7341 passing and 1 failing out of 7342 test cases



Executed 87 out of 294 tests: 256 tests pass, 1 fails locally and 37 were skipped.

make: *** [bazel_coverage_test] Error 3

root: /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/execroot/__main__/bazel-out/k8-fastbuild/testlogs

complete to collect bazel result.

script returned exit code 2

```
hi @mjonss, can you telll me which is the actuall failed test case?

**Comment by Defined2014** [comment]
@jiyfhust please fix `TestPartitionRangeColumnsForExpr` test

```
=== RUN   TestPartitionRangeColumnsForExpr
    partition_pruning_test.go:529: 
        	Error Trace:	planner/core/partition_pruning_test.go:529
        	Error:      	Should be true
        	Test:       	TestPartitionRangeColumnsForExpr
        	Messages:   	unexpected: a = 4 AND c = 3 [{1 10}] != [{1 9}]
```

**Review by Defined2014 — APPROVED** [review]

**Comment by Defined2014** [comment]
@mjonss Because of CI updated, seems need a re-approve for this PR.

**Review by xhebox — APPROVED** [review]

**Pull request merged**